staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go

```diff
@@ -36,12 +38,9 @@ func MaxLength(_ operation.Context, fldPath *field.Path, value, _ *string, max i
 
 // Required verifies that the specified value is not the zero-value for its
 // type.
-func Required[T comparable](_ operation.Context, fldPath *field.Path, value, _ *T) field.ErrorList {
```

Comment:
@jpbetz any opinions on changing this from `comparable` to `any`?  As @yongruilin noted it seems that there are currently `required` and `optional` usages in `pkg/apis/core/v1` on fields that are not comparable but want to understand a bit why `comparable` was used initially.  Thanks!

Reply:
Required was introduced by https://github.com/jpbetz/kubernetes/commit/adccb90f0319b6fa84aac3452020453f85f0b9f7.

I don't love having to switch to reflection here, but I don't know how else to support slices and maps.

cc @thockin

Reply:
What does it mean for a slice to be required, optional, or forbidden?  

optional == minItems=0
required == minItems=1
forbidden == maxItems=0

Should we just change the tags to say that?  We don't generally distinguish an empty slice from a nil slice (I am not sure the unmarshalling handles it correctly in all cases) so `sliceField: []` and `sliceField: null` and `// no sliceField` are all equivalent, right?

Alternately:
```diff
diff --git a/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go b/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go
index cadeb53681f..6495b1db302 100644
--- a/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go
+++ b/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go
@@ -98,7 +98,9 @@ const (
 )
 
 var (
-       optionalValidator = types.Name{Package: libValidationPkg, Name: "Optional"}
+       optionalValidator      = types.Name{Package: libValidationPkg, Name: "Optional"}
+       optionalSliceValidator = types.Name{Package: libValidationPkg, Name: "OptionalSlice"}
+       optionalMapValidator   = types.Name{Package: libValidationPkg, Name: "OptionalMap"}
 )
 
 func (optionalDeclarativeValidator) ExtractValidations(t *types.Type, comments []string) (Validations, error) {
@@ -106,6 +108,15 @@ func (optionalDeclarativeValidator) ExtractValidations(t *types.Type, comments [
        if !optional {
                return Validations{}, nil
        }
+       for t.Kind == types.Alias {
+               t = t.Underlying
+       }
+       switch t.Kind {
+       case types.Slice:
+               return Validations{Functions: []FunctionGen{Function(optionalTagName, ShortCircuit|NonError, optionalSliceValidator)}}, nil
+       case types.Map:
+               return Validations{Functions: []FunctionGen{Function(optionalTagName, ShortCircuit|NonError, optionalMapValidator)}}, nil
+       }
        return Validations{Functions: []FunctionGen{Function(optionalTagName, ShortCircuit|NonError, optionalValidator)}}, nil
 }
 
```

Don't solve it in the callee, solve it in the caller!